src/dfx/src/commands/deps/pull.rs

```diff
@@ -0,0 +1,340 @@
+        let wasm_url = reqwest::Url::parse(&wasm_url_str)?;
+        write_to_tempfile_then_rename(wasm_url_str.as_bytes(), &wasm_url_txt_path)?;
+        pulled_canister.wasm_url = Some(wasm_url_str);
+
+        // download
+        let response = reqwest::get(wasm_url.clone()).await?;
+        let status = response.status();
+        if status.is_client_error() || status.is_server_error() {
+            bail!("Failed to download wasm from url: {wasm_url}.\n  StatusCode: {status}");
+        }
+        let content = response.bytes().await?;
+
+        // hash check
+        let mut hasher = Sha256::new();
+        hasher.update(&content);
+        let hash_download = hasher.finalize();
+        if hash_download.as_slice() != hash_on_chain {
+            bail!(
+                "Hash mismatch.
+on chain: {}
+download: {}",
+                hex::encode(hash_on_chain),
+                hex::encode(hash_download.as_slice())
+            );
+        }
+
+        write_to_tempfile_then_rename(&content, &wasm_path)?;
+    }
+    // get `candid:service` from downloaded wasm
+    let wasm = std::fs::read(&wasm_path)
+        .with_context(|| format!("Failed to read wasm from {:?}", &wasm_path))?;
+    let candid_service = get_metadata(&wasm, CANDID_SERVICE).with_context(|| {
+        format!(
+            "Failed to get {} metadata from {:?}",
+            CANDID_SERVICE, &wasm_path
+        )
+    })?;
+    let service_candid_path = get_service_candid_path(canister_id)?;
+    write_to_tempfile_then_rename(candid_service.as_bytes(), &service_candid_path)?;
+
+    // get `candid:args` from downloaded wasm
+    let candid_args = get_metadata(&wasm, CANDID_ARGS).with_context(|| {
+        format!(
+            "Failed to get {} metadata from {:?}",
+            CANDID_ARGS, &wasm_path
+        )
+    })?;
+    pulled_canister.candid_args = Some(candid_args);
+
+    // try get `dfx:init` from downloaded wasm
+    match get_metadata(&wasm, DFX_INIT) {
+        Ok(dfx_init) => pulled_canister.dfx_init = Some(dfx_init),
+        Err(_e) => trace!(
+            logger,
+            "Failed to get {} metadata from {:?}",
+            DFX_INIT,
+            &wasm_path
+        ),
+    }
+
+    Ok(())
+}
+
+#[context("Failed to fetch metadata {metadata} of canister {canister_id}.")]
+async fn fetch_metatdata(
+    agent: &Agent,
+    canister_id: Principal,
+    metadata: &str,
+) -> DfxResult<Option<Vec<u8>>> {
+    match agent
+        .read_state_canister_metadata(canister_id, metadata)
+        .await
+    {
+        Ok(data) => Ok(Some(data)),
+        Err(agent_error) => match agent_error {
+            // replica returns such error
+            AgentError::HttpError(ref e) => {
+                let content = String::from_utf8(e.content.clone())?;
+                if content.starts_with("Custom section") {
+                    Ok(None)
+                } else {
+                    bail!(agent_error);
+                }
+            }
```

Comment:
The replica returns 404 in this case, which is pretty intuitive.  I would suggest both of the following:
1. check for status 404
2. check for `.starts_with("Custom section {metadata} not found")`
```
The replica returned an HTTP Error: Http Error: status 404 Not Found, content type "application/cbor", content: Custom section a:b:c not found.
```